beacon-chain/rpc/eth/beacon/handlers_validator.go

```diff
@@ -0,0 +1,375 @@
+				containers[i] = valContainerFromReadOnlyVal(val, primitives.ValidatorIndex(i), allBalances[i], valStatus)
+			} else {
+				containers[i] = valContainerFromReadOnlyVal(val, ids[i], allBalances[ids[i]], valStatus)
+			}
+		}
+		resp := &GetValidatorsResponse{
+			Data:                containers,
+			ExecutionOptimistic: isOptimistic,
+			Finalized:           isFinalized,
+		}
+		http2.WriteJson(w, resp)
+		return
+	}
+
+	filteredStatuses := make(map[validator.ValidatorStatus]bool, len(statuses))
+	for _, ss := range statuses {
+		ok, vs := validator.ValidatorStatusFromString(ss)
+		if !ok {
+			http2.HandleError(w, "Invalid status "+ss, http.StatusBadRequest)
+			return
+		}
+		filteredStatuses[vs] = true
+	}
+	valContainers := make([]*ValidatorContainer, 0, len(readOnlyVals))
+	for i, val := range readOnlyVals {
+		valStatus, err := helpers.ValidatorStatus(val, epoch)
+		if err != nil {
+			http2.HandleError(w, "Could not get validator status: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		valSubStatus, err := helpers.ValidatorSubStatus(val, epoch)
+		if err != nil {
+			http2.HandleError(w, "Could not get validator status: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
+		if filteredStatuses[valStatus] || filteredStatuses[valSubStatus] {
+			var container *ValidatorContainer
+			if len(ids) == 0 {
+				container = valContainerFromReadOnlyVal(val, primitives.ValidatorIndex(i), allBalances[i], valSubStatus)
+			} else {
+				container = valContainerFromReadOnlyVal(val, ids[i], allBalances[ids[i]], valSubStatus)
+			}
+			valContainers = append(valContainers, container)
+		}
+	}
+
+	resp := &GetValidatorsResponse{
+		Data:                valContainers,
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+	}
+	http2.WriteJson(w, resp)
+}
+
+// GetValidator returns a validator specified by state and id or public key along with status and balance.
+func (s *Server) GetValidator(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "beacon.GetValidator")
+	defer span.End()
+
+	stateId := mux.Vars(r)["state_id"]
+	if stateId == "" {
+		http2.HandleError(w, "state_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+	valId := mux.Vars(r)["validator_id"]
+	if valId == "" {
+		http2.HandleError(w, "validator_id is required in URL params", http.StatusBadRequest)
+		return
+	}
+
+	st, err := s.Stater.State(ctx, []byte(stateId))
+	if err != nil {
+		shared.WriteStateFetchError(w, err)
+		return
+	}
+	ids, ok := decodeIds(w, st, []string{valId}, false /* ignore unknown */)
+	if !ok {
+		return
+	}
+	readOnlyVals, ok := valsFromIds(w, st, ids)
+	if !ok {
+		return
+	}
+	valSubStatus, err := helpers.ValidatorSubStatus(readOnlyVals[0], slots.ToEpoch(st.Slot()))
+	if err != nil {
+		http2.HandleError(w, "Could not get validator status: "+err.Error(), http.StatusInternalServerError)
+		return
+	}
+	bal, err := st.BalanceAtIndex(ids[0])
```

Comment:
this can panic if len ids == 0 I think from `valsFromIds`

Reply:
actually wondering when the `len(ids) == 0` actually gets triggeredin `valsFromIds` because line 165 
`ids, ok := decodeIds(w, st, []string{valId}, false /* ignore unknown */)` shows that something exists.